lightningd/htlc_set.c

```diff
@@ -0,0 +1,193 @@
+#include <common/memleak.h>
+#include <common/timeout.h>
+#include <lightningd/htlc_end.h>
+#include <lightningd/htlc_set.h>
+#include <lightningd/lightningd.h>
+#include <lightningd/peer_htlcs.h>
+
+/* Pre-declaration as fail_htlc_set refers to it and vice versa */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set);
+
+void fail_htlc_set(struct htlc_set *set, enum onion_type failcode)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
+		fail_htlc(set->htlcs[i], failcode);
+	}
+	tal_free(set);
+}
+
+/* If an HTLC times out, we need to free entire set, since we could be processing
+ * it in invoice.c right now. */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set)
+{
+	/* For non-EXPERIMENTAL_FEATURES, this never actually called. */
+#if EXPERIMENTAL_FEATURES
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		if (set->htlcs[i] == hin) {
+			/* Don't try to re-fail this HTLC! */
+			tal_arr_remove(&set->htlcs, i);
+			/* Kind of the correct failure code. */
+			fail_htlc_set(set, WIRE_MPP_TIMEOUT);
+			return;
+		}
+	}
+	abort();
+#endif
+}
+
+void fulfill_htlc_set(struct htlc_set *set, const struct preimage *preimage)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
+		fulfill_htlc(set->htlcs[i], preimage);
+	}
+	tal_free(set);
+}
+
+#if EXPERIMENTAL_FEATURES
+static void destroy_htlc_set(struct htlc_set *set,
+			     struct htlc_set_map *map)
+{
+	htlc_set_map_del(map, set);
+}
+
+/* BOLT-9441a66faad63edc8cd89860b22fbf24a86f0dcd #4:
+ * - MUST fail all HTLCs in the HTLC set after some reasonable
+ *   timeout.
+ *   - SHOULD use `mpp_timeout` for the failure message.
+ */
+static void timeout_htlc_set(struct htlc_set *set)
+{
+	fail_htlc_set(set, WIRE_MPP_TIMEOUT);
+}
+#endif /* EXPERIMENTAL_FEATURES */
+
+static struct htlc_set *new_htlc_set(struct lightningd *ld,
+				     struct htlc_in *hin,
+				     struct amount_msat total_msat)
+{
+	struct htlc_set *set;
+
+	set = tal(ld, struct htlc_set);
+	set->total_msat = set->so_far = total_msat;
```

Comment:
Why is `set->so_far` being set here, when it is also being set to `AMOUNT_MSAT(0)` two lines below?

Reply:
Oops, good catch.  I originally set it to the first payment, then realized the code was neater to start at 0.